src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -454,5 +460,68 @@ public void StopTimingWrite()
                 _writeTimingWrites--;
             }
         }
+
+        private IDisposable BeginConnectionScope()
+        {
+            if (Log.IsEnabled(LogLevel.Critical))
+            {
+                return Log.BeginScope(new ConnectionScope(ConnectionId));
+            }
+
+            return null;
+        }
+
+        private class ConnectionScope : IReadOnlyList<KeyValuePair<string, object>>
+        {
+            private readonly string _connectionId;
+
+            private string _cachedToString;
+
+            public ConnectionScope(string connectionId)
+            {
+                _connectionId = connectionId;
+            }
+
+            public KeyValuePair<string, object> this[int index]
+            {
+                get
+                {
+                    if (index == 0)
+                    {
+                        return new KeyValuePair<string, object>("ConnectionId", _connectionId);
+                    }
+
+                    throw new ArgumentOutOfRangeException(nameof(index));
+                }
+            }
+
+            public int Count => 1;
+
+            public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+            {
+                for (int i = 0; i < Count; ++i)
+                {
+                    yield return this[i];
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            public override string ToString()
+            {
+                if (_cachedToString == null)
+                {
+                    _cachedToString = string.Format(
+                        CultureInfo.InvariantCulture,
+                        "ConnectionId:{0}",
```

Comment:
How well do our [own loggers](https://github.com/aspnet/Logging) support scopes? e.g. console, Azure app services, etc. If these don't easily display the name of the scope along with the message, we should keep the connection id in the message for now.